File: backend/src/main/java/com/neritech/saas/ordemservico/dto/OSAdditionalModels.java
package com.neritech.saas.ordemservico.dto;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Future;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

public final class OSAdditionalModels {
    private OSAdditionalModels() {}

    public record ItemDraft(
            @NotBlank String operation,
            @NotBlank String itemType,
            Long sourceItemId,
            Long catalogItemId,
            @NotBlank @Size(max = 500) String description,
            @NotNull @Positive BigDecimal quantity,
            @Size(max = 20) String unit,
            BigDecimal amountDelta,
            Integer timeDeltaMinutes) {}

    public record CreateRequest(
            @NotBlank @Size(max = 160) String title,
            @NotBlank String reason,
            @NotEmpty List<@Valid ItemDraft> items) {}

    public record UpdateRequest(
            @NotBlank @Size(max = 160) String title,
            @NotBlank String reason,
            @NotEmpty List<@Valid ItemDraft> items) {}

    public record SubmitRequest(
            @NotBlank @Size(max = 160) String recipientName,
            @NotBlank @Size(max = 30) String channel,
            @NotBlank @Size(max = 180) String recipientMasked,
            @NotNull @Future LocalDateTime expiresAt) {}

    public record ItemDecision(
            @NotNull Long itemId,
            @NotBlank String decision,
            @Size(max = 500) String comment) {}

    public record PublicDecisionRequest(@NotEmpty List<@Valid ItemDecision> items) {}

    public record ItemResponse(
            Long id,
            String operation,
            String itemType,
            Long sourceItemId,
            Long catalogItemId,
            String description,
            BigDecimal quantity,
            String unit,
            BigDecimal amountDelta,
            Integer timeDeltaMinutes,
            String decision,
            String decisionComment) {}

    public record Response(
            Long id,
            Long ordemServicoId,
            Integer baseOsVersion,
            String title,
            String reason,
            String status,
            BigDecimal amountDelta,
            Integer timeDeltaMinutes,
            String recipientName,
            String recipientChannel,
            String recipientMasked,
            LocalDateTime tokenExpiresAt,
            LocalDateTime submittedAt,
            LocalDateTime viewedAt,
            LocalDateTime decidedAt,
            LocalDateTime revokedAt,
            Integer version,
            LocalDateTime createdAt,
            List<ItemResponse> items,
            List<String> allowedActions) {}

    /** Token bruto é efêmero: só é devolvido no comando submit. */
    public record SubmitResponse(Response request, String approvalToken) {}

    public record PublicResponse(
            Long requestId,
            String orderNumber,
            String title,
            String reason,
            String status,
            BigDecimal amountDelta,
            Integer timeDeltaMinutes,
            LocalDateTime expiresAt,
            List<ItemResponse> items) {}
}

File: FrontEnd/src/app/routes/os/additional-requests/os-additional.models.ts
export type OsAdditionalStatus =
  | 'RASCUNHO' | 'PRONTA_PARA_ENVIO' | 'PENDENTE' | 'VISUALIZADA'
  | 'APROVADA' | 'PARCIAL' | 'RECUSADA' | 'EXPIRADA' | 'REVOGADA'
  | 'SUBSTITUIDA' | 'CANCELADA';

export type OsAdditionalOperation = 'ADD' | 'UPDATE' | 'REMOVE';
export type OsAdditionalItemType = 'SERVICE' | 'PRODUCT' | 'OTHER';
export type OsAdditionalDecision = 'PENDING' | 'APPROVED' | 'REJECTED';

export interface OsAdditionalItemDraft {
  operation: OsAdditionalOperation;
  itemType: OsAdditionalItemType;
  sourceItemId?: number | null;
  catalogItemId?: number | null;
  description: string;
  quantity: number;
  unit?: string | null;
  amountDelta?: number | null;
  timeDeltaMinutes?: number | null;
}

export interface OsAdditionalCreateRequest {
  title: string;
  reason: string;
  items: OsAdditionalItemDraft[];
}

export interface OsAdditionalSubmitRequest {
  recipientName: string;
  channel: string;
  recipientMasked: string;
  expiresAt: string;
}

export interface OsAdditionalItem {
  id: number;
  operation: OsAdditionalOperation;
  itemType: OsAdditionalItemType;
  sourceItemId?: number | null;
  catalogItemId?: number | null;
  description: string;
  quantity: number;
  unit?: string | null;
  amountDelta: number;
  timeDeltaMinutes: number;
  decision: OsAdditionalDecision;
  decisionComment?: string | null;
}

export interface OsAdditionalRequest {
  id: number;
  ordemServicoId: number;
  baseOsVersion?: number | null;
  title: string;
  reason: string;
  status: OsAdditionalStatus;
  amountDelta: number;
  timeDeltaMinutes: number;
  recipientName?: string | null;
  recipientChannel?: string | null;
  recipientMasked?: string | null;
  tokenExpiresAt?: string | null;
  submittedAt?: string | null;
  viewedAt?: string | null;
  decidedAt?: string | null;
  revokedAt?: string | null;
  version?: number | null;
  createdAt?: string | null;
  items: OsAdditionalItem[];
  allowedActions: string[];
}

export interface OsAdditionalSubmitResponse {
  request: OsAdditionalRequest;
  approvalToken: string;
}

export interface OsAdditionalPublicRequest {
  requestId: number;
  orderNumber: string;
  title: string;
  reason: string;
  status: OsAdditionalStatus;
  amountDelta: number;
  timeDeltaMinutes: number;
  expiresAt?: string | null;
  items: OsAdditionalItem[];
}

export interface OsAdditionalPublicDecision {
  items: Array<{ itemId: number; decision: 'APPROVED' | 'REJECTED'; comment?: string | null }>;
}

export type OsAdditionalLoadState = 'idle' | 'loading' | 'ready' | 'empty' | 'forbidden' | 'error';

export const OS_ADDITIONAL_STATUS_LABELS: Record<OsAdditionalStatus, string> = {
  RASCUNHO: 'Rascunho', PRONTA_PARA_ENVIO: 'Pronta para envio', PENDENTE: 'Aguardando cliente',
  VISUALIZADA: 'Visualizada', APROVADA: 'Aprovada', PARCIAL: 'Aprovada parcialmente',
  RECUSADA: 'Recusada', EXPIRADA: 'Expirada', REVOGADA: 'Revogada', SUBSTITUIDA: 'Substituída', CANCELADA: 'Cancelada',
};
